ambry-store/src/main/java/com.github.ambry.store/Log.java

```diff
@@ -13,161 +13,304 @@
+        remainingUnallocatedCapacity.addAndGet(-segmentCapacity);
+      }
     }
-    logger.trace("Log : {} bytes written to the log from read channel at {}, bytesWritten: {}", file.getAbsolutePath(),
-        offset, bytesWritten);
+    activeSegment = segmentsByName.firstEntry().getValue();
   }
 
   /**
-   * Close this log
+   * Allocates a file named {@code filename} and of capacity {@code size}.
+   * @param filename the intended filename of the file.
+   * @param size the intended size of the file.
+   * @return a {@link File} instance that points to the created file named {@code filename} and capacity {@code size}.
+   * @throws IOException if the there is any I/O error in allocating the file.
    */
-  void close()
+  private File allocate(String filename, long size)
       throws IOException {
-    fileChannel.close();
+    // TODO (DiskManager changes): This is intended to "request" the segment file from the DiskManager which will have
+    // TODO (DiskManager changes): a pool of segments.
+    File segmentFile = new File(dataDir, filename);
+    Utils.preAllocateFileIfNeeded(segmentFile, size);
+    remainingUnallocatedCapacity.addAndGet(-size);
+    return segmentFile;
   }
 
-  public void flush()
+  /**
+   * Rolls the log segment over if required. Also ensures enough capacity.
+   * @param writeSize the size of the incoming write.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
+   *
+   */
+  private void rollOverIfRequired(long writeSize)
       throws IOException {
-    fileChannel.force(true);
+    if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
+      ensureCapacity(writeSize);
+      // this cannot be null since capacity has either been ensured or has thrown.
+      LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
+      logger.info("Rolling over writes to {} from {} on write of data of size {}. End offset was {} and capacity is {}",
+          nextActiveSegment.getName(), activeSegment.getName(), writeSize, activeSegment.getEndOffset(),
+          activeSegment.getCapacityInBytes());
+      activeSegment = nextActiveSegment;
+    }
   }
 
-  @Override
-  public void readInto(ByteBuffer buffer, long position)
+  /**
+   * Ensures that there is enough capacity of a write of size {@code writeSize}.
+   * @param writeSize the size of a subsequent write on the active log segment.
+   * @throws IllegalArgumentException if the {@code writeSize} is greater than a single segment's size
+   * @throws IllegalStateException if there no more capacity in the log.
+   * @throws IOException if any I/O error occurred as a part of ensuring capacity.
+   */
+  private void ensureCapacity(long writeSize)
```

Comment:
this method doesn't just ensure capacity alone. It also allocates next segment as well. Can we reflect that in method name. Ideally I would prefer to move the logic of allocating new segment into rollover method if you want to keep this method just to ensure capacity